src/Juvix/Compiler/Asm/Translation/FromCore.hs

```diff
@@ -252,9 +283,62 @@ genCode infoTable fi =
     snocPopTemp False code = DL.snoc code (mkInstr PopTemp)
     snocPopTemp True code = code
 
-    convertType :: Int -> Core.Type -> Type
-    convertType = unimplemented
-
 -- Be mindful that JuvixAsm types are explicitly uncurried, while
--- Core.Stripped types are always curried. If a function takes `n` arguments
+-- Core.Stripped types are always curried. If a function takes `n` arguments,
 -- then the first `n` arguments should be uncurried in its JuvixAsm type.
+convertType :: Int -> Core.Type -> Type
+convertType argsNum ty =
+  case ty of
+    Core.TyDynamic ->
+      TyDynamic
+    Core.TyPrim (Core.PrimInteger Core.PrimIntegerInfo {..}) ->
+      TyInteger (TypeInteger _infoMinValue _infoMaxValue)
+    Core.TyPrim (Core.PrimBool Core.PrimBoolInfo {..}) ->
+      TyBool (TypeBool _infoTrueTag _infoFalseTag)
+    Core.TyPrim Core.PrimString ->
+      TyString
+    Core.TyApp Core.TypeApp {..} ->
+      TyInductive (TypeInductive _typeAppSymbol)
+    Core.TyFun {} ->
+      let (tgt, tyargs) = Core.unfoldType ty
+          tyargs' = map convertNestedType tyargs
+          tgt' = convertType 0 tgt
+       in mkTypeFun (take argsNum tyargs') (mkTypeFun (drop argsNum tyargs') tgt')
+
+convertNestedType :: Core.Type -> Type
```

Comment:
we should have a brief comment explaining why we need a different conversion function for nested types